flaml/autogen/agent/user_proxy_agent.py

```diff
@@ -90,23 +94,96 @@ def _execute_code(self, code_blocks):
+        """Extract arguements as a dict from a string.
+        Args:
+            input_string: string to extract arguements from
+        Returns:
+            a dictionary or None
+        """
+
+        def _remove_newlines_outside_quotes(s):
+            """Remove newlines outside of quotes.
+
+            if calling json.loads(s), it will throw an error because of the newline in the query.
+            So this function removes the newline in the query outside of quotes.
+
+            Ex 1:
+            "{\n"tool": "python",\n"query": "print('hello')\nprint('world')"\n}" -> "{"tool": "python","query": "print('hello')\nprint('world')"}"
+            Ex 2:
+            "{\n  \"location\": \"Boston, MA\"\n}" -> "{"location": "Boston, MA"}"
+            """
+            result = []
+            inside_quotes = False
+            for c in s:
+                if c == '"':
+                    inside_quotes = not inside_quotes
+                if not inside_quotes and c == "\n":
+                    continue
+                if inside_quotes and c == "\n":
+                    c = "\\n"
+                if inside_quotes and c == "\t":
+                    c = "\\t"
+                result.append(c)
+            return "".join(result)
+
+        input_string = _remove_newlines_outside_quotes(input_string)
+        try:
+            args = json.loads(input_string)
+            return args
+        except json.JSONDecodeError:
+            return None
+
+    def _execute_function(self, func_call):
+        func_name = func_call.get("name", "")
+        func = self._functions.get(func_name, None)
+
+        is_exec_success = False
+        if func is not None:
+            arguments = self._extractArgs(func_call.get("arguments", ""))
+            if arguments is not None:
+                try:
+                    content = func(**arguments)
+                    is_exec_success = True
+                except Exception as e:
+                    content = f"Error: {e}"
+            else:
+                content = f"Error: Invalid arguments for function {func_name}."
+        else:
+            content = f"Error: Function {func_name} not found."
+
+        return is_exec_success, {
+            "name": func_name,
+            "role": "function",
+            "content": content,
+        }
+
     def auto_reply(self, message, sender, default_reply=""):
         """Generate an auto reply."""
-        code_blocks = extract_code(message)
+        if "function_call" in message:
+            is_exec_success, func_return = self._execute_function(message["function_call"])
+            self._send(func_return, sender)
+            return
+
+        code_blocks = extract_code(message["content"])
         if len(code_blocks) == 1 and code_blocks[0][0] == UNKNOWN:
             # no code block is found, lang should be `UNKNOWN``
-            self._send(default_reply, sender)
+            self._send({"role": "user", "content": default_reply}, sender)
```

Comment:
This change will potentially mislead the oai_message list. From this agent's point of view, this message is an "assistant" message.

Reply:
I am changing it back to send the message as str and add the role in _send function.